appserver/payara-appserver-modules/microprofile/metrics/src/main/java/fish/payara/microprofile/metrics/impl/ConcurrentGaugeImpl.java

```diff
@@ -62,25 +63,53 @@
     /**
      * Last point for which clearOld() has been run
      */
-    private Instant lastInstant = Instant.now();
-    
+    private Instant lastInstant = getCurrentMinute();
+
+    /**
+     * A map containing all the max count values that the count of the gauge has
+     * been at in the last and current minutes. The key is the time in minute
+     * that the value changed, the value is the max value of the count at the
+     * moment before it changed
+     */
+    private final ConcurrentMap<Instant, Long> maxCounts
+            = new ConcurrentHashMap<>();
+
     /**
-     * A map containing all the values that the count of the gauge has been at in the last minute.
-     * The key is the time in seconds that the value changed, the value is the value of the count at the moment
-     * before it changed
+     * A map containing all the min count values that the count of the gauge has
+     * been at in the last and current minutes. The key is the time in minute
+     * that the value changed, the value is the min value of the count at the
+     * moment before it changed
      */
-    private Map<Instant, Long> lastCounts = new ConcurrentHashMap<>();
-    
+    private final ConcurrentMap<Instant, Long> minCounts
+            = new ConcurrentHashMap<>();
+
```

Comment:
This is nice simplification, but when I understand correctly these maps should always have just single entry, since it's regularly cleared. So we might be ok just with simple variables.

Or due to concurrency, two `MinMax` instances - one for previously completed minute, and one for updating.

Reply:
I agree that the `Map` should be replaced with simple fields. It makes this appear more complicated than it is and also has overheads and higher risk of concurrency issues we can avoid. If I understand the API correctly 2 atomic longs and 2 volatile longs are a good way.
The atomics to count/set min/max of the not yet completed minute whenever inc/dec are called. If we see that the method is called when a new minute has completed just copy over the current value of the atomics to the volatile min/max fields for "last completed minute" before doing the atomic update. the getters can simply return the volatile min/max fields. No map or cleanup should be needed.